### PR TITLE
update: Don't report formulae that are moved within a tap but not renamed

### PR DESCRIPTION
**Pull request by jawshooah**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

It's misleading to report a formula as "new" when it's only been moved from one directory to another within the same tap.

Ref: [caskroom/homebrew-cask#22669 (comment)](https://github.com/caskroom/homebrew-cask/pull/22669#issuecomment-231457315)

### CODE REVIEW

File: Library/Homebrew/cmd/update-report.rb [hunk 1]
```diff
@@ -209,8 +209,12 @@ def report
...
         @report[:M] << tap.formula_file_to_name(src)
       when /^R\d{0,3}/
-        @report[:D] << tap.formula_file_to_name(src) if tap.formula_file?(src)
-        @report[:A] << tap.formula_file_to_name(dst) if tap.formula_file?(dst)
+        src_full_name = tap.formula_file_to_name(src)
+        dst_full_name = tap.formula_file_to_name(dst)
+        # Don't report formulae that are moved within a tap but not renamed
+        next if src_full_name == dst_full_name
+        @report[:D] << src_full_name if tap.formula_file?(src)
+        @report[:A] << dst_full_name if tap.formula_file?(dst)
```
Comment (UniqMartin) [inline Library/Homebrew/cmd/update-report.rb]: There's a very subtle problem with these two lines. If formulae happen to be moved within a tap *and* a formula gets renamed before or afterwards (thus `src_full_name != dst_full_name`) and the update covers both operations, then the report will look something like this (I experimented a bit with my local copy of the `homebrew/gui` tap):

```
Updated 1 tap (homebrew/gui).
==> New Formulae
homebrew/gui/xchat2
```

But it should have looked something like this:

```
Updated 1 tap (homebrew/gui).
==> Renamed Formulae
homebrew/gui/xchat -> homebrew/gui/xchat2
```

The reason is that `if tap.formula_file?(src)` will always be `false` as it checks against the current `Tap#formula_dir`. It's admittedly a rare corner case, but this should solve it:

```ruby
if tap.formula_file?(dst)
  @report[:D] << src_full_name
  @report[:A] << dst_full_name
end
```

**Comment by UniqMartin** [comment]
Thanks for taking a look and for this PR! I commented on a corner case, but otherwise it's looking good.

Unfortunately, other parts of Homebrew also assume that `Tap#formula_dir` never changes during the lifetime of a tap and thus break if formulae are relocated within a tap. See issue #87 for details. Do you maybe see yourself tackling these as well? (Would be awesome, but I certainly don't expect you to.)

File: Library/Homebrew/cmd/update-report.rb (hunk 1 above)
Comment (jawshooah) [inline Library/Homebrew/cmd/update-report.rb]: Nice catch! Fixed in 2648506

Comment (UniqMartin) [inline Library/Homebrew/cmd/update-report.rb]: Thanks for finding an even more elegant solution and particularly for adding a nice test case!

File: Library/Homebrew/cmd/update-report.rb [hunk 2]
```diff
@@ -209,8 +209,13 @@ def report
         end
         @report[:M] << tap.formula_file_to_name(src)
       when /^R\d{0,3}/
-        @report[:D] << tap.formula_file_to_name(src) if tap.formula_file?(src)
-        @report[:A] << tap.formula_file_to_name(dst) if tap.formula_file?(dst)
+        next unless tap.formula_file?(dst)
```
Comment (UniqMartin) [inline Library/Homebrew/cmd/update-report.rb]: I just realized this check can be dropped completely, as there's a check preceding the `case` statement that already makes sure that at least one of the modified paths is a formula file:

```ruby
next unless paths.any? { |p| tap.formula_file?(p) }
```

Comment (jawshooah) [inline Library/Homebrew/cmd/update-report.rb]: Got it, removed in 7195878.

**Comment by UniqMartin** [comment]
Merged in 4abdeb7b5eb06acb633a9ac12d7c681d736a5286. Thank you for this contribution to Homebrew, @jawshooah! :tada:

Sorry for the long delay! I'd be still interested in your answer to the following question:

> Unfortunately, other parts of Homebrew also assume that `Tap#formula_dir` never changes during the lifetime of a tap and thus break if formulae are relocated within a tap. See issue #87 for details. Do you maybe see yourself tackling these as well?

**Comment by jawshooah** [comment]
@UniqMartin No worries! Sorry for the delay in getting back to you. I haven't looked too deeply into `FormulaVersions` but it seems like a proper fix might require a deep dive into Git plumbing commands. I'll see about giving it a shot later this evening, but I can't make any promises.

**Pull request merged**